tracer_metrics.go

```diff
@@ -0,0 +1,369 @@
+		}),
+		connDurations: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "quic_connection_duration",
+			Help:    "Connection Duration",
+			Buckets: prometheus.ExponentialBuckets(1, 1.5, 40), // 1s to ~12 weeks
+		}),
+	}
+}
+
+var _ prometheus.Collector = &aggregatingCollector{}
+
+func (c *aggregatingCollector) Describe(descs chan<- *prometheus.Desc) {
+	descs <- c.rtts.Desc()
+	descs <- c.connDurations.Desc()
+}
+
+func (c *aggregatingCollector) Collect(metrics chan<- prometheus.Metric) {
+	now := time.Now()
+	c.mutex.Lock()
+	for _, conn := range c.conns {
+		if rtt, valid := conn.getSmoothedRTT(); valid {
+			c.rtts.Observe(rtt.Seconds())
+		}
+		c.connDurations.Observe(now.Sub(conn.startTime).Seconds())
+	}
+	c.mutex.Unlock()
+	metrics <- c.rtts
+	metrics <- c.connDurations
+}
+
+func (c *aggregatingCollector) AddConn(id string, t *metricsConnTracer) {
+	c.mutex.Lock()
+	c.conns[id] = t
+	c.mutex.Unlock()
+}
+
+func (c *aggregatingCollector) RemoveConn(id string) {
+	c.mutex.Lock()
+	delete(c.conns, id)
+	c.mutex.Unlock()
+}
+
+var collector *aggregatingCollector
+
+func init() {
+	const (
+		direction = "direction"
+		encLevel  = "encryption_level"
+	)
+
+	closedConns = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_connections_closed_total",
+			Help: "closed QUIC connection",
+		},
+		[]string{direction},
+	)
+	prometheus.MustRegister(closedConns)
+	newConns = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_connections_new_total",
+			Help: "new QUIC connection",
+		},
+		[]string{direction, "handshake_successful"},
+	)
+	prometheus.MustRegister(newConns)
+	bytesTransferred = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_transferred_bytes",
+			Help: "QUIC bytes transferred",
+		},
+		[]string{direction}, // TODO: this is confusing. Other times, we use direction for the perspective
+	)
+	prometheus.MustRegister(bytesTransferred)
+	sentPackets = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_packets_sent_total",
+			Help: "QUIC packets sent",
+		},
+		[]string{encLevel},
+	)
+	prometheus.MustRegister(sentPackets)
+	rcvdPackets = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_packets_rcvd_total",
+			Help: "QUIC packets received",
+		},
+		[]string{encLevel},
+	)
+	prometheus.MustRegister(rcvdPackets)
+	bufferedPackets = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_packets_buffered_total",
+			Help: "Buffered packets",
+		},
+		[]string{"packet_type"},
+	)
+	prometheus.MustRegister(bufferedPackets)
+	droppedPackets = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_packets_dropped_total",
+			Help: "Dropped packets",
+		},
+		[]string{"packet_type", "reason"},
+	)
+	prometheus.MustRegister(droppedPackets)
+	connErrors = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "quic_connection_errors",
```

Comment:
```suggestion
			Name: "quic_connection_errors_total",
```

Nit pick: This is a unit-less counter and should thus be suffixed with `_total`.

Reply:
Fixed.